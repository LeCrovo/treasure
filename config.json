{
  "apiPort": 8000,
  "uiPort": 3000,
  "defaultAlbum": "sample",
  "defaultGallery": "demo",
  "resizeDimensions": {
    "photo": { "width": 800, "height": 600 },
    "thumb": { "width": 185, "height": 45 }
  },
  "supportedFileTypes": {
    "photo": ["jpg", "jpeg"],
    "video": ["mp4", "webm"]
  },
  "rawFileTypes": {
    "photo": ["arw", "raw"],
    "video": ["avi", "mov", "m2ts", "mts"]
  }
}
