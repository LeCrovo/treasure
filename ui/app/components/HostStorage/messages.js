import { defineMessages } from 'react-intl';

export const scope = 'app.components.HostStorage';

export default defineMessages({
  instruction: {
    id: `${scope}.instruction`,
    defaultMessage: '{host} token/path for loading gallery/media',
  },
});
