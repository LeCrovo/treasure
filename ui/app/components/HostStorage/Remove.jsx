import React, { memo, useEffect } from 'react';
import { FormattedMessage } from 'react-intl';

import { getHostToken, hostCase } from '../../utils/host';
import { remove as removeFromStorage } from '../../utils/localStorage';

import Button from '../Button';
import GenericList from '../GenericList/Loadable';
import H3 from '../H3';
import ListItem from '../ListItem';
import messages from './messages';

function Remove({
  showHeader = true,
  setTokenInStorage = () => {},
  storageUpdated = () => {},
}) {
  const storedHostnames = ['dropbox', 'local'].filter(host =>
    getHostToken(host, 'browser'),
  );

  useEffect(() => {
    if (storedHostnames.length > 0) {
      setTokenInStorage(true);
    }
  });

  function handleRemove(host) {
    removeFromStorage(host);
    storageUpdated(host, null, false);
  }

  const RemoveButtons = ({ item: { host } }) => (
    <ListItem
      item={
        <Button onClick={() => handleRemove(host)}>
          {hostCase(host)}
          <span role="img" aria-label="Delete token" title="Delete token">
            ❌️
          </span>
        </Button>
      }
    />
  );

  return (
    <section>
      {showHeader && (
        <H3>
          <FormattedMessage {...messages.clearHeader} />
        </H3>
      )}
      <GenericList
        loading={false}
        error={false}
        items={storedHostnames.map(host => ({ id: `remove-${host}`, host }))}
        component={RemoveButtons}
      />
    </section>
  );
}

export default memo(Remove);
