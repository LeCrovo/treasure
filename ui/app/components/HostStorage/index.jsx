import { Formik, Field, Form } from 'formik';
import React, { memo, useState } from 'react';
import styled from 'styled-components';

import { FormattedMessage } from 'react-intl';

import { hostCase } from '../../utils/strings';
import messages from './messages';

const ErrorMessage = styled.span`
  color: red;
`;

function isUrl(value) {
  const pattern = /(http|https):\/\/[\w-]+(\.[\w-]+)+([\w.,@?^=%&amp;:/~+#-]*[\w@?^=%&amp;/~+#-])?/g;
  return new RegExp(pattern).test(value);
}

function isToken(value) {
  const pattern = /^[0-9a-z_]{10,}$/gi;
  return new RegExp(pattern).test(value);
}

function HostStorage({ host }) {
  const [stored, setStored] = useState(false);
  const hostName = hostCase(host);

  function handleTokenValidation(value) {
    const isValid = isUrl(value) || isToken(value);

    if (isValid) {
      return null;
    }

    setStored(false);
    return 'Invalid token (min 10 char) or URL';
  }

  function handleSubmit({ token }) {
    localStorage.setItem(`host-${host}`, token);
    setStored(true);
  }

  return (
    <Formik initialValues={{ token: '' }} onSubmit={handleSubmit}>
      {({ errors, touched, isValidating }) => (
        <Form>
          <FormattedMessage
            {...messages.instruction}
            values={{
              host: hostName,
            }}
          />
          <Field type="text" name="token" validate={handleTokenValidation} />
          {errors.token && touched.token && (
            <ErrorMessage>{errors.token}</ErrorMessage>
          )}
          <button type="submit" disabled={isValidating}>
            Store in browser
          </button>
          {stored && (
            <span role="img" aria-label="success">
              âœ…
            </span>
          )}
        </Form>
      )}
    </Formik>
  );
}

export default memo(HostStorage);
