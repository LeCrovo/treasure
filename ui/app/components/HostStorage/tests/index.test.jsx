import React from 'react';
import { fireEvent, render, act } from '@testing-library/react';
import { IntlProvider } from 'react-intl';

import HostStorage from '../index';
import { apiPort } from '../../../../../config.json';
import { DEFAULT_LOCALE } from '../../../locales';

describe('<HostStorage />', () => {
  test('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <HostStorage />
      </IntlProvider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  test('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <HostStorage />
      </IntlProvider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  test('Form validation should prevent blank', async () => {
    const { getByLabelText, getByTestId } = render(<HostStorage host="cdn" />);

    await act(async () => {
      fireEvent.change(getByLabelText('Token input'), {
        target: { value: '' },
      });
    });

    await act(async () => {
      fireEvent.submit(getByTestId('form'));
    });

    expect(getByLabelText('Token error').innerHTML).toBe(
      'Invalid token (min 10 char) or URL',
    );
  });

  test('Should allow a localhost URL', async () => {
    const { getByLabelText, getByTestId } = render(<HostStorage host="cdn" />);

    await act(async () => {
      fireEvent.change(getByLabelText('Token input'), {
        target: { value: `http://localhost:${apiPort}` },
      });
    });

    await act(async () => {
      fireEvent.submit(getByTestId('form'));
    });

    expect(getByLabelText('success').innerHTML).toBe('âœ…');
  });
});
