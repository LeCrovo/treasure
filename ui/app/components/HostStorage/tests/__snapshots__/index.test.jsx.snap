// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HostStorage /> Should render and match the snapshot 1`] = `
<form
  action="#"
  data-testid="form"
>
  <span>
     token/path for loading gallery/media
  </span>
  <input
    aria-label="Token input"
    name="token"
    type="text"
    value=""
  />
  <input
    aria-label="Submit button"
    type="submit"
    value="Store in browser"
  />
</form>
`;
